File: boj/python/1717.py
def find(x):
    if x == parent[x]:
        return x
    parent[x] = find(parent[x])
    return parent[x]

def union(x, y):
    x = find(x)
    y = find(y)
    if x == y:
        return
    if x < y:
        parent[y] = x
    else:
        parent[x] = y

def solve(k, a, b):
    if k == 0:
        union(a, b)
    else:
        if a == b or find(a) == find(b):
            print("YES")
        else:
            print("NO")

n, m = map(int, input().split())
parent = [i for i in range(n + 1)]

for _ in range(m):
    k, a, b = map(int, input().split())
    solve(k, a, b)

File: boj/cpp/1717.cpp
#include<bits/stdc++.h>
using namespace std;

int n, m, k, a, b;
int parent[1000002];

int Find(int x) {
	if (x == parent[x]) return x;

	return parent[x] = Find(parent[x]);
}

void Union(int x, int y) {
	x = Find(x);
	y = Find(y);

	if (x == y) return;

	if (x < y) parent[y] = x;
	else parent[x] = y;
}

void solve(int k, int a, int b){
    if(k == 0){
        Union(a, b);        
    } else {
        if(a == b || Find(a) == Find(b)) cout << "YES\n";
        else cout << "NO\n";
    }
}

int main(void){
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> m;
    for(int i = 0; i <= n; i++) parent[i] = i;

    while(m--){
        cin >> k >> a >> b;
        solve(k, a, b);
    }
    return 0;
}